rust-runtime/aws-smithy-http-server/src/request.rs

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Parse HTTP request trait
+
+use aws_smithy_http::body::SdkBody;
+use bytes::Bytes;
+
+/// `ParseHttpRequest` is a generic trait for parsing structured data from HTTP requests.
+///
+/// It is designed to be flexible, because `Input` is unconstrained, it can be used to support
+/// event streams, regular request-response style operations, as well as any other HTTP-based
+/// protocol that we manage to come up with.
+///
+/// It also enables this critical and core trait to avoid being async, and it makes code that uses
+/// the trait easier to test.
+pub trait ParseHttpRequest {
+    /// Input type of the HttpRequest
+    ///
+    /// For request/response style operations, this is typically something like:
+    /// `Result<OperationInput, OperationInputError>`
```

Comment:
If this already returns an operation input, what do we need `axum`'s `FromRequest` for? Isn't this already the extractor?

Also, won't the error type be fixed? I'd assume the error type to be an enum with generic errors that can occur during parsing that are **not** specific to the particular operation input being extracted, e.g. things like `axum`'s rejection types: https://docs.rs/axum/0.2.8/axum/extract/rejection/index.html

Reply:
1) I thought about using directly axum traits, but it was easier for me to approach this change by first creating my own simple traits. This allowed me to understand how the code generation works for operation I/O. Once this PR and the one implementing the codegen are landed, I think we should probably switch to the axum traits.

2) The error here should be just the generic server Error type, I'll fix the documentation.